lib/carrierwave/uploader/versions.rb

```diff
@@ -213,23 +213,60 @@ def url(*args)
       # Recreate versions and reprocess them. This can be used to recreate
       # versions if their parameters somehow have changed.
       #
-      def recreate_versions!(*versions)
+      def recreate_versions!(*names)
         # Some files could possibly not be stored on the local disk. This
         # doesn't play nicely with processing. Make sure that we're only
         # processing a cached file
         #
         # The call to store! will trigger the necessary callbacks to both
         # process this version and all sub-versions
-        if versions.any?
-          file = sanitized_file if !cached?
-          store_versions!(file, versions)
+
+        if names.any?
+          file = sanitized_file
+          set_versions_to_cache_and_store(names)
+          cache!(file)
+          store!(file)
+          reset_versions_to_cache_and_store
         else
           cache! if !cached?
           store!
         end
       end
 
     private
+
+      def set_versions_to_cache_and_store(names)
+        @versions_to_cache = source_versions_of(names)
```

Comment:
Basically, it's still an instance variable. And now even two of them instead of one 😄 . But introducing them helps to avoid conditions in `cache_versions!` and `store_versions!` methods. Also, assigning these variables is hidden in methods with readable names. So, I think this change decreases overall complexity.